### PR TITLE
Add `Crystal::System::Addrinfo`

### PR DESCRIPTION
**Pull request by HertzDevil**
Moves the platform-specific code into a separate module, so that implementations other than `LibC.getaddrinfo` can be added without cluttering the same file (e.g. Win32's `GetAddrInfoExW` from #13619). Depends on #14956.

### CODE REVIEW

**Review by ysbaddaden — COMMENTED** [review]
What about moving the yielding `getaddrinfo(&)` as a system method? That would abstract the next/free logic.

File: src/crystal/system/addrinfo.cr [hunk 1]
```diff
@@ -0,0 +1,39 @@
+module Crystal::System::Addrinfo
+  # alias Handle
+
+  # protected def initialize(addrinfo : Handle)
+
+  # def system_ip_address : ::Socket::IPAddress
+
+  # only used by `#system_ip_address`?
+  # def to_unsafe
```
Comment (straight-shoota) [inline src/crystal/system/addrinfo.cr:9]: **question:** Why is `to_unsafe` part of the interface? It seems to be used only internally in the implementations.

**Comment by straight-shoota** [comment]
@ysbaddaden Maybe `getaddrinfo(&)` should actually be pushed down entirely to the implementation level. Both current implementations work by iterating abstract handles. But that doesn't have to be the case for all implementations.
It's fine to leave it as is for now, though. We can adapt when the need arises.

**Comment by straight-shoota** [comment]
**thought:** I'm wondering if it really makes sense to split the `getaddrinfo` implementations for Unix and Windows.
The function is almost identical between POSIX and WinSock. So we have a lot of duplication in the handling code. IMO the merged code was actually quite readable with some target-specific conditionals here and there.

**Comment by HertzDevil** [comment]
> The function is almost identical between POSIX and WinSock

Don't worry, `GetAddrInfoExW` will look different enough

File: src/crystal/system/addrinfo.cr (hunk 1 above)
Comment (HertzDevil) [reply src/crystal/system/addrinfo.cr:9]: It has to be somewhere, I don't think we could outright remove it even if we expand the internal use of it.

**Comment by straight-shoota** [comment]
So the plan is to switch entirely to `GetAddrInfoExW` in a follow-up? Sounds great, then!

**Comment by ysbaddaden** [comment]
My intent was for the system API to be the yielding `getaddrinfo(&)` and skip the next/free methods + passing an object (internal détail). Each system implementation will be free to iterate the resolved addresses as they need.

File: src/crystal/system/addrinfo.cr (hunk 1 above)
Comment (straight-shoota) [reply src/crystal/system/addrinfo.cr:9]: No, I mean we can keep the actual `def to_unsafe` but drop the commented pseudo-interface definition here. That's only intended for methods that are called externally.

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**